Optionally shuffle operations prior to execution pass.

csadorf opened the pull request:
## Description

Optionally shuffle operations prior to execution pass with the `run --shuffle` command line flag or `shuffle` argument to `FlowProject.run()`.

## Motivation and Context

Shuffling may serve multiple purposes:

 * Speed up execution for parallelized runs, where operations that
   operate on the same job may block eachother.
 * Simplify debugging since errors that may occur later in the workflow
   are more likely to occur a bit earlier.

## Types of Changes
<!-- Please select all items that apply either now or after creating the pull request: -->
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:
<!-- Please select all items that apply either now or after creating the pull request. -->
<!-- If you are unsure about any of these items, do not hesitate to ask! -->
- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

csadorf commented:
Alternatively we could just shuffle by default, what do you think?

b-butler commented:
With groups coming up, I am not sure of the best choice. Shuffling by default seems like a good option with the caveat it may require that more use cases pass a `--num-passes=2` option to the group.

vyasr commented:
I'm fine with using this randomization, but we should probably also change the default operation order to sort things by operation to maximize the chances that different operations on the same job will not be run in parallel. That change should also minimize any parallel file access issues that occur when multiple operations acting on the same job execute simultaneously.

b-butler commented:
@vyasr I think that is a good idea as well.

csadorf commented:
I've adjusted the implementation to actually be a bit more flexible, I think the only remaining question is whether we want to change the default behavior. Possible downside is just the overhead for rearrangement, but I'd opt for making 'round-robin' the default.

b-butler commented on flow/project.py:
```diff
@@ -1491,11 +1491,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
+                * callable specify a key-callable by which to order operations
+            The default value is None.
         """
+        from itertools import groupby
```
I am not sure why the imports are going here. If the purpose is conditional importing, shouldn't it be further down in the if statements? Otherwise, I think it should go to the top with the other imports as is typical python form.

b-butler commented on flow/project.py:
```diff
@@ -1439,7 +1439,7 @@ def _fork(self, operation, timeout=None):
 
     def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=None,
-            num_passes=1, progress=False, shuffle=False):
+            num_passes=1, progress=False, order=None):
```
You say in a comment that job ordering is done by default. Unless that changes, I think the default value should be `'by_job'`. Strings are not mutable in python so this would be a fine default value, and would more clear specify to users in the documentation that default ordering is `by-job`.

csadorf replied:
Making the default value 'by-job' instead of `None` means that we are guaranteeing a specific order. If ever wanted to change that, then we need to go through a whole deprecation cycle. It's much safer to make no guarantees about the order by default to avoid future incompatibilities. The fact that `none` and `by-job` happen to be the same one for our specific implementation is therefore essentially of no relevance with respect to the default value to this argument.

There is one more reason why I'd prefer to have `None` be the default: Our execution model is generally based on the idea of a state-machine, where the definition order and execution order of operations should in principle not matter. The only reason I am introducing this option in the first place is to allow to speed up the execution under specific circumstances.

b-butler replied:
That makes sense. I just didn't understand the reasoning before.

b-butler commented on flow/project.py:
```diff
@@ -1570,8 +1576,23 @@ def select(operation):
+            elif order == 'random':
                 random.shuffle(operations)
+            elif order is None or order in ('none', 'by-job'):
+                pass  # by-job is the default order
```
See comment above.

b-butler commented on flow/project.py:
```diff
@@ -2584,9 +2605,11 @@ class MyProject(FlowProject):
+            '--order',
+            type=str,
+            choices=['none', 'by-job', 'round-robin', 'random'],
+            default=None,
```
I think default should be `'by-job'` for previously specified reasons.

csadorf commented:
> I've adjusted the implementation to actually be a bit more flexible, I think the only remaining question is whether we want to change the default behavior. Possible downside is just the overhead for rearrangement, but I'd opt for making 'round-robin' the default.

Having thought about this a bit longer, I am now advising to leave the default oder at `none`, i.e., no specific order, reasons are outlined in my comment to @b-butler abover.

bdice commented on flow/project.py:
```diff
@@ -1490,7 +1491,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
             Show a progress bar during execution.
         :type progess:
             bool
+        :param order:
```
Does it get a type docstring, `str`?

bdice commented on flow/project.py:
```diff
@@ -1490,7 +1491,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
             bool
+        :param order:
+            Specify the order of operations, possible values are:
+                * None or 'none' (no reordering, do not assume any order)
```
I'd prefer to specify what this does instead of what it does not do. Something like "uses order as returned by (some function)" or some definition from the FlowProject. I'm not sure how this works internally, but I find the current docstring unclear and would want to search the source code for clarification of what "no assumed order" actually means.

csadorf replied:
Like I explained in my response to @b-butler , I'd prefer to not explain what the code does "implicitly" in order to avoid users starting to rely on that. I think it would be highly preferable for us to have the option to change the default behavior at any point in time without needing to go through a deprecation cycle.

This is similar to the debate about Python's built-in `dict` class, where the insertion order was considered implementation detail up until Python 3.7, although it has been identical to `OrderedDict` for quite a while now. See this [StackOverflow reply](https://stackoverflow.com/a/39980744) for some details. I don't think it is a good idea for us to promise any kind of order while the implementation is still highly in flux.

bdice replied:
I guess I didn’t realize this was the same as @b-butler’s suggestion. As a new user I would have difficulty understanding the default behavior as it is. I think it would be fine to make by-job the default and even change the execution order without a full deprecation, if you view the API guarantee as the end result of the state machine instead of the exact order of execution. However, I understand your reasoning and am fine with your decisions here.

bdice commented on flow/project.py:
```diff
@@ -1490,7 +1491,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
+        :param order:
+            Specify the order of operations, possible values are:
+                * None or 'none' (no reordering, do not assume any order)
+                * 'by-job'  (operations are grouped by job)
```
Spacing of these lines is weird - it looks like they were meant to be aligned at one point. I recommend keeping it consistent at one space between the string `'by-job'` and the description `(operations are grouped by job)`, for all elements of this list.

csadorf replied:
Done.

bdice commented on flow/project.py:
```diff
@@ -1490,7 +1491,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
+            Specify the order of operations, possible values are:
+                * None or 'none' (no reordering, do not assume any order)
+                * 'by-job'  (operations are grouped by job)
+                * 'round-robin' (order operations in round-robin fashion by job)
```
Alternatively this could be called 'cyclic' -- to me, 'round-robin' _could_ suggest a finite time slot offered to each operation, as described here: https://en.wikipedia.org/wiki/Round-robin_scheduling

It's clear enough as-is, just wanted to offer this as a suggestion in case you like it better.

csadorf replied:
I've switched it to cyclic.

bdice commented on flow/project.py:
```diff
@@ -1490,7 +1491,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
+                * 'by-job'  (operations are grouped by job)
+                * 'round-robin' (order operations in round-robin fashion by job)
+                * 'random'    (shuffle the execution order randomly)
+                * callable specify a key-callable by which to order operations
```
Would prefer `* callable (specify a key-callable used to sort operations)`

bdice commented on flow/project.py:
```diff
@@ -1565,6 +1576,23 @@ def select(operation):
+            else:
+                raise ValueError(
+                    "Invalid value for 'order' argument, valid arguments are "
+                    "None, 'none', 'by-job', 'round-robin', or 'random'.")
```
or callable?

bdice commented on flow/project.py:
```diff
@@ -1565,6 +1576,23 @@ def select(operation):
+                operations = list(roundrobin(*groups))
+            elif order == 'random':
+                random.shuffle(operations)
+            elif order is None or order in ('none', 'by-job'):
```
Why allow both `None` and `'none'`? Seems unnecessary and increases the API scope.

csadorf replied:
The reason for this is so that we can easily specify 'None' from the command line, consider the following example:
```#!bash
for order in none round-robin random; do
    ./project.py run --order=$order -o foo bar
done
```
If `none` was not a valid argument, one would need to drop the `--order` argument altogether, which means you need to either make the code more verbose:
```#!bash
for order in --order=none --order=round-robin --order=random; do
    ./project.py run $order -o foo bar
done
```
or add an awkward special case with potential code duplication:
```#!bash
./project.py run foo bar
for order in round-robin random; do
    ./project.py run --order=$order -o foo bar
done
```
I've seen this kind of pattern before and would argue that that little bit of redundancy is warranted. If it helps, we can make `none` the documented default argument and avoid the redundancy as part of the API documentation.

bdice replied:
Good reasoning, I am fine with it this way now that I understand.

bdice commented on flow/project.py:
```diff
@@ -2575,6 +2604,12 @@ class MyProject(FlowProject):
             const='-1',
             help="Specify the number of cores to parallelize to. Defaults to all available "
                  "processing units if argument is ommitted.")
```
fix typo: `omitted`

bdice commented on flow/util/misc.py:
```diff
@@ -199,3 +200,19 @@ def keys_used(self):
+    "roundrobin('ABC', 'D', 'EF') --> A D E B F C"
+    # Recipe credited to George Sakkis
+    num_active = len(iterables)
+    nexts = cycle(iter(it).__next__ for it in iterables)
```
See? It's cyclic. (See previous comment) 😉

csadorf replied:
:D I'll change that.

bdice commented on flow/util/misc.py:
```diff
@@ -199,3 +200,19 @@ def keys_used(self):
+
+def roundrobin(*iterables):
+    # From: https://docs.python.org/3/library/itertools.html#itertools-recipes
+    "roundrobin('ABC', 'D', 'EF') --> A D E B F C"
```
Would prefer a `# comment` instead of a string literal.

csadorf commented:
@bdice I think I've addressed all of your comments and suggestions.
@b-butler I've tried to clarify my motivation for the choice of default argument both in a comment to you and have also tried to clarify this as part of the API documentation.

@vyasr @b-butler I'm still open to changing the default behavior, but I would like the default argument to remain 'none'.

vyasr commented on flow/project.py:
```diff
@@ -1490,6 +1493,26 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
+                * 'by-job' (operations are grouped by job)
+                * 'cyclic' (order operations cyclic by job)
+                * 'random' (explicitly shuffle the execution order randomly)
+                * callable (specify a key-callable used to sort operations)
```
This description isn't too clear to me, mostly because `key-callable` isn't really a canonical term AFAIK. I would prefer to word this as something like `a callable returning a comparison key for an operation that is used to sort operations`, that matches the terminology used in the documentation of `sorted` (comparison key).

vyasr commented on flow/project.py:
```diff
@@ -1490,6 +1493,26 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
+                * 'none' or None (no specific order)
+                * 'by-job' (operations are grouped by job)
+                * 'cyclic' (order operations cyclic by job)
+                * 'random' (explicitly shuffle the execution order randomly)
```
no need for "explicitly"

vyasr commented on flow/project.py:
```diff
@@ -1490,6 +1493,26 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
+            .. note::
+
+                Users are advised to not rely on a specific execution order, as a
+                subsitute for defining the worklow in terms of pre- and post-conditions.
```
typos: "subsitute"->"substitute", "worklow"->"workflow"

vyasr commented on flow/util/misc.py:
```diff
@@ -199,3 +200,19 @@ def keys_used(self):
+    # roundrobin('ABC', 'D', 'EF') --> A D E B F C
+    # Recipe credited to George Sakkis
+    num_active = len(iterables)
+    nexts = cycle(iter(it).__next__ for it in iterables)
```
Are you inclined to leave this recipe unchanged since it was copied? If so, that's fine, but if you're open to changes I would rewrite this as `cycle(iter(next(it)))` rather than calling `__next__`.

csadorf replied:
Your solution would actually be wrong (I tested it). So I'm leaving as-is.

vyasr replied:
Interesting. Do you know why that is?

csadorf replied:
It is because we are creating a cycle of `__next__` functions, not a cycle of *next* elements.

vyasr commented:
> @bdice I think I've addressed all of your comments and suggestions.
> @b-butler I've tried to clarify my motivation for the choice of default argument both in a comment to you and have also tried to clarify this as part of the API documentation.
> 
> @vyasr @b-butler I'm still open to changing the default behavior, but I would like the default argument to remain 'none'.

I'm fine leaving the default behavior as is for now, we can revisit later if needed.

csadorf commented:
@vyasr I've addressed all of your change requests.

vyasr commented on flow/project.py:
```diff
@@ -1498,16 +1498,17 @@ def run(self, jobs=None, names=None, pretend=False, np=None, timeout=None, num=N
-                * 'random' (explicitly shuffle the execution order randomly)
-                * callable (specify a key-callable used to sort operations)
+                * 'random' (shuffle the execution order randomly)
+                * callable (a callable returning a a comparison key for an
```
typo: "a a comparison key"

csadorf replied:
damn :D

csadorf commented:
@bdice @vyasr @b-butler Thank you for helping with this feature!!
